### PR TITLE
fix: support masked_scatter by lowering path and corner case of maske…

### PR DESCRIPTION
**Pull request by chohk88**
# Description

Implemented support for `masked_scatter` in the lowering path, referring to [this implementation in PyTorch Inductor](https://github.com/pytorch/pytorch/blob/924a247fbbe32063966533535b1cc0cf474ef1ed/torch/_inductor/decomposition.py#L831).

Fixes # ([issue](https://github.com/pytorch/TensorRT/issues/3410))

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/lowering/_decompositions.py [hunk 1]
```diff
@@ -575,6 +577,53 @@ def cudnn_grid_sampler_decomposition(
...
+    gathered = source_flat.gather(0, safe_idx)
+
+    # 5) Replace masked positions in input with gathered values
+    replaced = torch.where(mask_flat, gathered, input_flat)
+
+    # 6) Reshape the result to match the original broadcasted shape
+    return replaced.view(input_b.shape)
+
+
 def get_decompositions(
```
Comment (apbose) [inline py/torch_tensorrt/dynamo/lowering/_decompositions.py:627]: @chohk88 the implementation looks perfect as such.'
Just a slight detail I was wondering. Torch mentions that it supports broadcasting between mask and self tensor, but I see this example working too, basically broadcasting between source and mask.

import torch
input = torch.zeros(3, 3)
mask = torch.tensor([[1, 0, 1], [0, 0, 1], [0, 0, 0]], dtype=torch.bool)
source = torch.tensor([2, 3, 4])
out = input.masked_scatter_(mask, source)
print("out is", out)

Comment (chohk88) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:627]: Thanks @apbose, great question! I checked PyTorch’s behavior and it:

1. Broadcasts only input and mask (not source)
2. Flattens source as-is
3. Uses only the first mask.sum() elements of source, ignoring any extras

For example, with source=[2,3,4,5] and three True mask positions, PyTorch applies [2,3,4] and drops the 5 without error.

Our current decomposition does exactly that (broadcasts input/mask, flattens source, then cumsum→gather→where), so no code changes are needed. Let me know if you spot any other edge cases!

image
image
image

Comment (apbose) [reply py/torch_tensorrt/dynamo/lowering/_decompositions.py:627]: great! Thanks for the detailed explanation
